llvm/lib/ProfileData/InstrProf.cpp

```diff
@@ -821,15 +822,23 @@ void InstrProfRecord::merge(InstrProfRecord &Other, uint64_t Weight,
 
   for (size_t I = 0, E = Other.Counts.size(); I < E; ++I) {
     bool Overflowed;
-    uint64_t Value =
-        SaturatingMultiplyAdd(Other.Counts[I], Weight, Counts[I], &Overflowed);
-    if (Value > getInstrMaxCountValue()) {
-      Value = getInstrMaxCountValue();
-      Overflowed = true;
+    uint64_t Value;
+    // When a profile has single byte coverage, use || to merge counters.
+    if (HasSingleByteCoverage)
+      Value = Other.Counts[I] || Counts[I];
```

Comment:
As I understand, this is needed so that when we merge multiple raw coverage profiles we get an indexed profile with counters that are either 0 or 1.
If we instead merge by adding counters, we would get technically inaccurate counts. However, they would be accurate if interpreted for coverage, i.e., 0 is not covered and >0 is covered. In addition, indexed profiles would have more information because if a counter is large, we know it is commonly executed because it was covered in many raw profiles.
Do you think it makes sense to merge these as counters for the indexed profile and simply interpret them as coverage?

Reply:
Yes, this is to make sure that we only write 0s or 1s into indexed profiles when we use single byte counters mode.  

re: "In addition, indexed profiles would have more information because if a counter is large, we know it is commonly executed because it was covered in many raw profiles.": Do you have a use case for this? Unless there is a use case, I prefer the values in raw profiles and indexed profiles to be consistent. Otherwise, it might be cause confusion and difficulty in debugging. Currently, we only explored single byte counter in raw profiles. If there is any benefit, we can also explore using single byte counters in the indexed profiles.

Reply:
Yes, we actually do use block coverage to collect many raw profiles and aggregate them into an indexed profile. Since the counts can be larger than one, it helps differentiate between some hot and cold blocks during optimization.

Reply:
I'm actually surprised this doesn't break any tests. If not, then I might need to add some more.

Reply:
Ok, there is a valid use case then. So, I removed this code.